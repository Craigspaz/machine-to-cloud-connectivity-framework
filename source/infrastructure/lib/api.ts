// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Aws, Construct, RemovalPolicy } from '@aws-cdk/core';
import { ServicePrincipal } from '@aws-cdk/aws-iam';
import { Function as LambdaFunction } from '@aws-cdk/aws-lambda';
import {
  AccessLogFormat,
  AuthorizationType,
  ContentHandling,
  Deployment,
  EndpointType,
  Integration,
  IntegrationType,
  LogGroupLogDestination,
  MethodLoggingLevel,
  MethodOptions,
  PassthroughBehavior,
  RequestValidator,
  RestApi,
  Stage
} from '@aws-cdk/aws-apigateway';
import { LogGroup, RetentionDays } from '@aws-cdk/aws-logs';
import { addCfnSuppressRules } from '../utils/utils';

/**
 * ApiConstructProps props
 * @interface ApiConstructProps
 */
export interface ApiConstructProps {
  // Connection builder Lambda function
  readonly connectionBuilderLambdaFunction: LambdaFunction;
}

/**
 * @class
 * Machine to Cloud Connectivity Framework API Construct.
 * It creates an API Gateway REST API and other resources.
 */
export class ApiConstruct extends Construct {
  // API endpoint
  public apiEndpoint: string;
  // API ID
  public apiId: string;

  constructor(scope: Construct, id: string, props: ApiConstructProps) {
    super(scope, id);

    const { connectionBuilderLambdaFunction } = props;
    const connectionBuilderLambdaFunctionArn = connectionBuilderLambdaFunction.functionArn;

    const apiLogGroup = new LogGroup(this, 'Logs', {
      removalPolicy: RemovalPolicy.DESTROY,
      retention: RetentionDays.THREE_MONTHS
    });
    addCfnSuppressRules(apiLogGroup, [{
      id: 'W84', reason: 'CloudWatch Logs are already encrypted by default.'
    }]);

    const api = new RestApi(this, 'M2C2Api', {
      defaultCorsPreflightOptions: {
        allowOrigins: ['*'],
        allowHeaders: [
          'Authorization',
          'Content-Type',
          'X-Amz-Date',
          'X-Amz-Security-Token',
          'X-Api-Key'
        ],
        allowMethods: [
          'GET',
          'POST',
          'OPTIONS'
        ],
        statusCode: 200
      },
      deploy: true,
      deployOptions: {
        accessLogDestination: new LogGroupLogDestination(apiLogGroup),
        accessLogFormat: AccessLogFormat.jsonWithStandardFields(),
        loggingLevel: MethodLoggingLevel.INFO,
        stageName: 'prod',
        tracingEnabled: true
      },
      description: 'Machine to Cloud Connectivity Rest API',
      endpointTypes: [EndpointType.REGIONAL]
    });
    this.apiEndpoint = `https://${api.restApiId}.execute-api.${Aws.REGION}.amazonaws.com/prod`;
    this.apiId = api.restApiId;

    const requestValidator = new RequestValidator(this, 'ApiRequestValidator', {
      restApi: api,
      validateRequestParameters: true,
      validateRequestBody: true
    });

    addCfnSuppressRules(api.node.findChild('Deployment') as Deployment, [{
      id: 'W68', reason: 'The solution does not require the usage plan.'
    }]);
    addCfnSuppressRules(api.node.findChild('DeploymentStage.prod') as Stage, [{
      id: 'W64', reason: 'The solution does not require the usage plan.'
    }]);

    const getAllIntegration = new Integration({
      type: IntegrationType.AWS_PROXY,
      integrationHttpMethod: 'POST',
      options: {
        contentHandling: ContentHandling.CONVERT_TO_TEXT,
        integrationResponses: [{ statusCode: '200' }],
        passthroughBehavior: PassthroughBehavior.WHEN_NO_MATCH,
        requestParameters: { 'integration.request.querystring.nextToken': 'method.request.querystring.nextToken' }
      },
      uri: `arn:${Aws.PARTITION}:apigateway:${Aws.REGION}:lambda:path/2015-03-31/functions/${connectionBuilderLambdaFunctionArn}/invocations`
    });
    const getAllMethodOptions: MethodOptions = {
      authorizationType: AuthorizationType.IAM,
      methodResponses: [{
        statusCode: '200',
        responseModels: {
          'application/json': { modelId: 'Empty' }
        }
      }],
      requestParameters: { 'method.request.querystring.nextToken': false },
      requestValidator: requestValidator
    };

    /**
     * Connections API
     * GET /connections
     * POST /connections
     * GET /connections/{connectionName}
     */
    const connectionsResource = api.root.addResource('connections');
    connectionsResource.addMethod(
      'POST',
      new Integration({
        type: IntegrationType.AWS_PROXY,
        integrationHttpMethod: 'POST',
        options: {
          contentHandling: ContentHandling.CONVERT_TO_TEXT,
          integrationResponses: [{ statusCode: '200' }],
          passthroughBehavior: PassthroughBehavior.WHEN_NO_MATCH
        },
        uri: `arn:${Aws.PARTITION}:apigateway:${Aws.REGION}:lambda:path/2015-03-31/functions/${connectionBuilderLambdaFunctionArn}/invocations`
      }),
      {
        authorizationType: AuthorizationType.IAM,
        methodResponses: [{
          statusCode: '200',
          responseModels: {
            'application/json': { modelId: 'Empty' }
          }
        }],
        requestValidator: requestValidator
      }
    );
    connectionsResource.addMethod('GET', getAllIntegration, getAllMethodOptions);

    const connectionResource = connectionsResource.addResource('{connectionName}');
    connectionResource.addMethod(
      'GET',
      new Integration({
        type: IntegrationType.AWS_PROXY,
        integrationHttpMethod: 'POST',
        options: {
          contentHandling: ContentHandling.CONVERT_TO_TEXT,
          integrationResponses: [{ statusCode: '200' }],
          passthroughBehavior: PassthroughBehavior.WHEN_NO_MATCH,
          requestParameters: { 'integration.request.path.connectionName': 'method.request.path.connectionName' }
        },
        uri: `arn:${Aws.PARTITION}:apigateway:${Aws.REGION}:lambda:path/2015-03-31/functions/${connectionBuilderLambdaFunctionArn}/invocations`
      }),
      {
        authorizationType: AuthorizationType.IAM,
        methodResponses: [{
          statusCode: '200',
          responseModels: {
            'application/json': { modelId: 'Empty' }
          }
        }],
        requestParameters: { 'method.request.path.connectionName': true },
        requestValidator: requestValidator
      }
    );

    /**
     * Logs API
     * GET /logs
     * GET /logs/{connectionName}
     */
    const logsResource = api.root.addResource('logs');
    logsResource.addMethod('GET', getAllIntegration, getAllMethodOptions);
    logsResource.addResource('{connectionName}').addMethod(
      'GET',
      new Integration({
        type: IntegrationType.AWS_PROXY,
        integrationHttpMethod: 'POST',
        options: {
          contentHandling: ContentHandling.CONVERT_TO_TEXT,
          integrationResponses: [{ statusCode: '200' }],
          passthroughBehavior: PassthroughBehavior.WHEN_NO_MATCH,
          requestParameters: {
            'integration.request.path.connectionName': 'method.request.path.connectionName',
            'integration.request.querystring.nextToken': 'method.request.querystring.nextToken'
          }
        },
        uri: `arn:${Aws.PARTITION}:apigateway:${Aws.REGION}:lambda:path/2015-03-31/functions/${connectionBuilderLambdaFunctionArn}/invocations`
      }),
      {
        authorizationType: AuthorizationType.IAM,
        methodResponses: [{
          statusCode: '200',
          responseModels: {
            'application/json': { modelId: 'Empty' }
          }
        }],
        requestParameters: {
          'method.request.path.connectionName': true,
          'method.request.querystring.nextToken': false
        },
        requestValidator: requestValidator
      }
    );

    /**
     * Sitewise API
     * GET /sitewise/{serverName}
     */
     const sitewiseResource = api.root.addResource('sitewise');
     sitewiseResource.addResource('{serverName}').addMethod(
       'GET',
       new Integration({
         type: IntegrationType.AWS_PROXY,
         integrationHttpMethod: 'POST',
         options: {
           contentHandling: ContentHandling.CONVERT_TO_TEXT,
           integrationResponses: [{ statusCode: '200' }],
           passthroughBehavior: PassthroughBehavior.WHEN_NO_MATCH,
           requestParameters: {
             'integration.request.path.serverName': 'method.request.path.serverName'
           }
         },
         uri: `arn:${Aws.PARTITION}:apigateway:${Aws.REGION}:lambda:path/2015-03-31/functions/${connectionBuilderLambdaFunctionArn}/invocations`
       }),
       {
         authorizationType: AuthorizationType.IAM,
         methodResponses: [{
           statusCode: '200',
           responseModels: {
             'application/json': { modelId: 'Empty' }
           }
         }],
         requestParameters: {
           'method.request.path.serverName': true
         },
         requestValidator: requestValidator
       }
     );

    connectionBuilderLambdaFunction.addPermission('ApiLambdaInvokePermission', {
      action: 'lambda:InvokeFunction',
      principal: new ServicePrincipal('apigateway.amazonaws.com'),
      sourceArn: api.arnForExecuteApi()
    });
  }
}