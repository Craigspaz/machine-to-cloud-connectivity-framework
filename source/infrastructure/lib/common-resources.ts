// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Aws, Construct, RemovalPolicy } from '@aws-cdk/core';
import { Effect, Policy, PolicyStatement, ServicePrincipal } from '@aws-cdk/aws-iam';
import { BlockPublicAccess, Bucket, BucketAccessControl, BucketEncryption, CfnBucket } from '@aws-cdk/aws-s3';

/**
 * Machine to Cloud Connectivity Framework Common Resources
 */
export class CommonResourcesConstruct extends Construct {
  // CloudWatch Logs policy
  private cloudWatchLogsPolicy: Policy;
  // S3 Logging bucket
  private s3LoggingBucket: Bucket;
  // S3 Resource Storage bucket
  private s3Bucket: Bucket;

  constructor(scope: Construct, id: string) {
    super(scope, id);

    this.cloudWatchLogsPolicy = new Policy(this, 'CloudWatchLogsPolicy', {
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: [
            'logs:CreateLogGroup',
            'logs:CreateLogStream',
            'logs:PutLogEvents'
          ],
          resources: [
            `arn:${Aws.PARTITION}:logs:${Aws.REGION}:${Aws.ACCOUNT_ID}:log-group:/aws/lambda/*`
          ]
        })
      ]
    });

    this.s3LoggingBucket = new Bucket(this, 'LogBucket', {
      accessControl: BucketAccessControl.LOG_DELIVERY_WRITE,
      blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
      encryption: BucketEncryption.S3_MANAGED,
      removalPolicy: RemovalPolicy.RETAIN
    });
    const cfnBucket = this.s3LoggingBucket.node.defaultChild as CfnBucket;
    cfnBucket.addMetadata('cfn_nag', {
      rules_to_suppress: [
        { id: 'W35', reason: 'This bucket is to store S3 logs, so it does not require access logs.' },
        { id: 'W51', reason: 'This bucket is to store S3 logs, so it does not require S3 policy.' }
      ]
    });

    this.s3Bucket = new Bucket(this, 'M2C2Bucket', {
      blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
      encryption: BucketEncryption.S3_MANAGED,
      removalPolicy: RemovalPolicy.RETAIN,
      serverAccessLogsBucket: this.s3LoggingBucket,
      serverAccessLogsPrefix: 'm2c2/'
    });
    this.s3Bucket.addToResourcePolicy(
      new PolicyStatement({
        effect: Effect.ALLOW,
        principals: [ new ServicePrincipal('lambda.amazonaws.com') ],
        actions: [
          's3:GetObject',
          's3:PutObject'
        ],
        resources: [
          `${this.s3Bucket.bucketArn}/*`
        ]
      })
    );
  }

  /**
   * Get CloudWatch Logs policy.
   * @return {Policy} CloudWatch Logs policy
   */
  getCloudWatchLogsPolicy(): Policy {
    return this.cloudWatchLogsPolicy;
  }

  /**
   * Get S3 logging bucket.
   * @return {Bucket} Logging bucket
   */
  getS3LoggingBucket(): Bucket {
    return this.s3LoggingBucket;
  }

  /**
   * Get S3 M2C2 bucket.
   * @return {Bucket} M2C2 bucket
   */
  getS3Bucket(): Bucket {
    return this.s3Bucket;
  }

  /**
   * Get S3 M2C2 bucket name
   * @return {string} M2C2 bucket name
   */
  getS3BucketName(): string {
    return this.s3Bucket.bucketName;
  }
}