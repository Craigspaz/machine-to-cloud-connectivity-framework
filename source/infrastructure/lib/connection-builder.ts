// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Aws, CfnDeletionPolicy, Construct, Duration } from '@aws-cdk/core';
import { AttributeType, CfnTable } from '@aws-cdk/aws-dynamodb';
import {
  Effect,
  Policy,
  PolicyDocument,
  PolicyStatement,
  Role,
  ServicePrincipal
} from '@aws-cdk/aws-iam';
import { Code, Function as LambdaFunction, Runtime } from '@aws-cdk/aws-lambda';
import { IBucket } from '@aws-cdk/aws-s3';
import { LambdaToDynamoDB } from '@aws-solutions-constructs/aws-lambda-dynamodb';
import { addCfnSuppressRules } from '../utils/utils';
import { CfnGateway } from '@aws-cdk/aws-iotsitewise';

/**
 * ConnectionBuilderConstruct props
 * @interface ConnectionBuilderConstructProps
 */
export interface ConnectionBuilderConstructProps {
  // Policy for CloudWatch Logs
  readonly cloudWatchLogsPolicy: Policy;
  // IoT endpoint address
  readonly iotEndpointAddress: string;
  // IoT Sitewise gateway
  readonly iotSitewiseGateway: CfnGateway;
  // Greengrass group ID
  readonly greengrassGroupId: string;
  // Kinesis Data Stream name
  readonly kinesisStreamName: string;
  // Logs DynamoDB table ARN
  readonly logsTableArn: string;
  // Logs DynamoDB table name
  readonly logsTableName: string;
  /**
   * Solution config properties.
   * Logging level, solution ID, version, source code bucket, and source code prefix
   */
  readonly solutionConfig: {
    loggingLevel: string;
    sendAnonymousUsage: string;
    solutionId: string;
    solutionVersion: string;
    sourceCodeBucket: IBucket;
    sourceCodePrefix: string;
  };
  // Solution UUID
  readonly uuid: string;
}

/**
 * @class
 * Machine to Cloud Connectivity Framework Connection Builder Construct.
 * It creates a connection builder Lambda function, a connection metadata DynamoDB table, an M2C2 S3 bucket, an IoT rule, and an API Gateway.
 */
export class ConnectionBuilderConstruct extends Construct {
  // Connection metadata DynamoDB table name
  public connectionDynamodbTableName: string;
  // Connection builder Lambda function
  public connectionBuilderLambdaFunction: LambdaFunction;

  constructor(scope: Construct, id: string, props: ConnectionBuilderConstructProps) {
    super(scope, id);

    const sourceCodeBucket = props.solutionConfig.sourceCodeBucket;
    const sourceCodePrefix = props.solutionConfig.sourceCodePrefix;
    const iotSitewiseGatewayArn = `arn:${Aws.PARTITION}:iotsitewise:${Aws.REGION}:${Aws.ACCOUNT_ID}:gateway/${props.iotSitewiseGateway.ref}`;

    const greengrassLambdaRole = new Role(this, 'GreengrassLambdaRole', {
      assumedBy: new ServicePrincipal('lambda.amazonaws.com'),
      path: '/'
    });
    greengrassLambdaRole.attachInlinePolicy(props.cloudWatchLogsPolicy);

    const lambdaToDynamoDb = new LambdaToDynamoDB(this, 'ConnectionBuilder', {
      lambdaFunctionProps: {
        code: Code.fromBucket(sourceCodeBucket, `${sourceCodePrefix}/connection-builder.zip`),
        description: 'Machine to Cloud Connectivity connection builder function',
        environment: {
          IOT_ENDPOINT: props.iotEndpointAddress,
          IOT_SITEWISE_GATEWAY_ID: props.iotSitewiseGateway.ref,
          KINESIS_STREAM: props.kinesisStreamName,
          LOGGING_LEVEL: props.solutionConfig.loggingLevel,
          LOGS_DYNAMODB_TABLE: props.logsTableName,
          PAGE_SIZE: '20',
          SEND_ANONYMOUS_METRIC: props.solutionConfig.sendAnonymousUsage,
          SOLUTION_ID: props.solutionConfig.solutionId,
          SOLUTION_VERSION: props.solutionConfig.solutionVersion,
          SOLUTION_UUID: props.uuid
        },
        handler: 'connection-builder/index.handler',
        runtime: Runtime.NODEJS_14_X,
        timeout: Duration.minutes(1)
      },
      dynamoTableProps: {
        partitionKey: {
          name: 'connectionName',
          type: AttributeType.STRING
        }
      },
      tableEnvironmentVariableName: 'CONNECTION_DYNAMODB_TABLE'
    });
    this.connectionBuilderLambdaFunction = lambdaToDynamoDb.lambdaFunction;

    const cfnDynamoTable = lambdaToDynamoDb.dynamoTable.node.defaultChild as CfnTable;
    cfnDynamoTable.cfnOptions.deletionPolicy = CfnDeletionPolicy.DELETE;
    this.connectionDynamodbTableName = lambdaToDynamoDb.dynamoTable.tableName;

    const greengrassDeployerRole = new Role(this, 'GreengrassDeployerRole', {
      assumedBy: new ServicePrincipal('lambda.amazonaws.com'),
      path: '/',
      inlinePolicies: {
        'GreengrassIoTPolicy': new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: [
                'greengrass:Create*DefinitionVersion',
                'greengrass:Create*Definition',
                'greengrass:CreateDeployment',
                'greengrass:CreateGroupVersion',
                'greengrass:Get*DefinitionVersion',
                'greengrass:GetDeploymentStatus',
                'greengrass:GetGroupVersion',
                'greengrass:GetGroup',
                'greengrass:List*Definitions'
              ],
              resources: ['*']
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ['iot:Publish'],
              resources: [
                `arn:${Aws.PARTITION}:iot:${Aws.REGION}:${Aws.ACCOUNT_ID}:topic/m2c2/job/*`,
                `arn:${Aws.PARTITION}:iot:${Aws.REGION}:${Aws.ACCOUNT_ID}:topic/m2c2/info/*`,
                `arn:${Aws.PARTITION}:iot:${Aws.REGION}:${Aws.ACCOUNT_ID}:topic/m2c2/error/*`
              ]
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: [
                'iotsitewise:DescribeGatewayCapabilityConfiguration',
                'iotsitewise:UpdateGatewayCapabilityConfiguration'
              ],
              resources: [iotSitewiseGatewayArn]
            })
          ]
        }),
        'LambdaPolicy': new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: [
                'lambda:CreateFunction',
                'lambda:DeleteFunction',
                'lambda:PublishVersion',
                'lambda:CreateAlias',
                'lambda:DeleteAlias'
              ],
              resources: [`arn:${Aws.PARTITION}:lambda:${Aws.REGION}:${Aws.ACCOUNT_ID}:function:m2c2-*`]
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ['iam:PassRole'],
              resources: [greengrassLambdaRole.roleArn]
            })
          ]
        }),
        'DynamoDBPolicy': new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: [
                'dynamodb:DeleteItem',
                'dynamodb:GetItem',
                'dynamodb:PutItem',
                'dynamodb:Scan',
                'dynamodb:UpdateItem',
                'dynamodb:Query',
              ],
              resources: [lambdaToDynamoDb.dynamoTable.tableArn]
            })
          ]
        }),
        'S3Policy': new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ['s3:GetObject'],
              resources: [`arn:${Aws.PARTITION}:s3:::${sourceCodeBucket.bucketName}/${sourceCodePrefix}/*`]
            })
          ]
        })
      }
    });
    greengrassDeployerRole.attachInlinePolicy(props.cloudWatchLogsPolicy);
    addCfnSuppressRules(greengrassDeployerRole, [{
      id: 'W11', reason: 'The * resource is needed to control Greengrass resources.'
    }]);

    const greengrassDeployer = new LambdaFunction(this, 'GreengrassDeployer', {
      code: Code.fromBucket(sourceCodeBucket, `${sourceCodePrefix}/greengrass-deployer.zip`),
      description: 'Machine to Cloud Connectivity Greengrass deployer function',
      environment: {
        LAMBDA_ROLE: greengrassLambdaRole.roleArn,
        GREENGRASS_GROUP_ID: props.greengrassGroupId,
        IOT_ENDPOINT: props.iotEndpointAddress,
        IOT_SITEWISE_GATEWAY_ID: props.iotSitewiseGateway.ref,
        CONNECTION_DYNAMODB_TABLE: this.connectionDynamodbTableName,
        KINESIS_STREAM: props.kinesisStreamName,
        LOGGING_LEVEL: props.solutionConfig.loggingLevel,
        SEND_ANONYMOUS_METRIC: props.solutionConfig.sendAnonymousUsage,
        SOLUTION_ID: props.solutionConfig.solutionId,
        SOLUTION_VERSION: props.solutionConfig.solutionVersion,
        SOLUTION_UUID: props.uuid,
        SOURCE_S3_BUCKET: sourceCodeBucket.bucketName,
        SOURCE_S3_PREFIX: sourceCodePrefix
      },
      handler: 'greengrass-deployer/index.handler',
      retryAttempts: 0,
      reservedConcurrentExecutions: 1,
      role: greengrassDeployerRole,
      runtime: Runtime.NODEJS_14_X,
      timeout: Duration.minutes(5)
    });
    this.connectionBuilderLambdaFunction.addEnvironment('GREENGRASS_DEPLOYER_LAMBDA_FUNCTION', greengrassDeployer.functionName);

    const connectionBuilderPolicy = new Policy(this, 'ConnectionBuilderPolicy', {
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: ['iot:Publish'],
          resources: [
            `arn:${Aws.PARTITION}:iot:${Aws.REGION}:${Aws.ACCOUNT_ID}:topic/m2c2/job/*`,
            `arn:${Aws.PARTITION}:iot:${Aws.REGION}:${Aws.ACCOUNT_ID}:topic/m2c2/info/*`
          ]
        }),
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: ['lambda:InvokeFunction'],
          resources: [greengrassDeployer.functionArn]
        }),
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: [
            'dynamodb:Scan',
            'dynamodb:Query'
          ],
          resources: [props.logsTableArn]
        }),
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: [
            'iotsitewise:DescribeGatewayCapabilityConfiguration',
            'iotsitewise:UpdateGatewayCapabilityConfiguration'
          ],
          resources: [iotSitewiseGatewayArn]
        })
      ]
    });

    const connectionBuilderRole = this.connectionBuilderLambdaFunction.role as Role;
    connectionBuilderRole.attachInlinePolicy(connectionBuilderPolicy);
  }
}