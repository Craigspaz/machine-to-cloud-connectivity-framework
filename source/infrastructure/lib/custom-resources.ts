// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import {
  Aws,
  CfnCondition,
  CfnCustomResource,
  CfnMapping,
  Construct,
  CustomResource,
  Duration
} from '@aws-cdk/core';
import { Policy, Role, ServicePrincipal } from '@aws-cdk/aws-iam';
import { CfnFunction, Code, Function, Runtime } from '@aws-cdk/aws-lambda';
import { Bucket } from '@aws-cdk/aws-s3';

/**
 * CustomResourcesConstruct props
 */
export interface CustomResourcesConstructProps {
  /**
   * Policy for CloudWatch Logs
   */
  readonly cloudWatchLogsPolicy: Policy;
  /**
   * Existing Greengrass group ID
   */
  readonly existingGreengrassGroup: string;
  /**
   * Existing Kinesis Data Stream name
   */
  readonly existingKinesisStream: string;
  /**
   * Condition for sending anonymous usage
   */
  readonly sendAnonymousUsageCondition: CfnCondition;
  /**
   * Mapping for the source code
   */
  readonly sourceCodeMapping: CfnMapping;
  /**
   * Mapping for the solution information
   */
  readonly solutionMapping: CfnMapping;
}

/**
 * Machine to Cloud Connectivity Framework Custom Resources
 */
export class CustomResourcesConstruct extends Construct {
  // Solution UUID
  private uuid: string;

  constructor(scope: Construct, id: string, props: CustomResourcesConstructProps) {
    super(scope, id);

    const helperFunctionRole = new Role(this, 'HelperFunctionRole', {
      assumedBy: new ServicePrincipal('lambda.amazonaws.com'),
      path: '/'
    });
    helperFunctionRole.attachInlinePolicy(props.cloudWatchLogsPolicy);

    const helperFunction = new Function(this, 'HelperFunction', {
      description: 'Machine to Cloud Connectivity custom resource function',
      handler: 'm2c2_helper_custom_resource.handler',
      runtime: Runtime.PYTHON_3_8,
      code: Code.fromBucket(
        Bucket.fromBucketName(this, 'SourceCodeBucket', `${props.sourceCodeMapping.findInMap('General', 'S3Bucket')}-${Aws.REGION}`),
        `${props.sourceCodeMapping.findInMap('General', 'KeyPrefix')}/m2c2-helper.zip`
      ),
      timeout: Duration.seconds(240),
      role: helperFunctionRole,
      environment: {
        SOLUTION_ID: props.solutionMapping.findInMap('Parameters', 'Id'),
        SOLUTION_VERSION: props.solutionMapping.findInMap('Parameters', 'Version'),
      }
    });
    const cfnFunction = helperFunction.node.defaultChild as CfnFunction;
    cfnFunction.addMetadata('cfn_nag', {
      rules_to_suppress: [
        { id: 'W58', reason: 'Miss alarm, and the function does have permission to write CloudWatch Logs.' }
      ]
    });

    const customUuid = new CustomResource(this, 'UUID', {
      serviceToken: helperFunction.functionArn,
      properties: {
        Resource: 'CreateUUID'
      }
    });
    this.uuid = customUuid.getAtt('UUID').toString();

    const sendAnonymousMetrics = new CustomResource(this, 'SendAnonymousMetrics', {
      serviceToken: helperFunction.functionArn,
      properties: {
        Resource: 'SendAnonymousMetrics',
        SolutionId: props.solutionMapping.findInMap('Parameters', 'Id'),
        UUID: this.uuid,
        Version: props.solutionMapping.findInMap('Parameters', 'Version'),
        Region: Aws.REGION,
        ExistingGreengrassGroup: props.existingGreengrassGroup,
        ExistingKinesisStream: props.existingGreengrassGroup
      }
    });
    const cfnSendAnonymousMetrics = sendAnonymousMetrics.node.defaultChild as CfnCustomResource;
    cfnSendAnonymousMetrics.cfnOptions.condition = props.sendAnonymousUsageCondition;
  }

  /**
   * Get solution UUID.
   * @return {string} UUID
   */
  getUuid(): string {
    return this.uuid;
  }
}