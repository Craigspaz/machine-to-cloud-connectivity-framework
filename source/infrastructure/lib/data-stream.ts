// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Construct } from '@aws-cdk/core';
import { Bucket } from '@aws-cdk/aws-s3';
import { KinesisStreamsToKinesisFirehoseToS3 } from '@aws-solutions-constructs/aws-kinesisstreams-kinesisfirehose-s3';

/**
 * DataStreamConstruct props
 */
export interface DataStreamConstructProps {
  /**
   * Logging S3 Bucket
   */
  readonly s3LoggingBucket: Bucket;
}

/**
 * Machine to Cloud Connectivity Framework Kinesis Streams, Kinesis Firehose, and S3
 */
export class DataStreamConstruct extends Construct {
  // Kinesis Data Stream name
  private kinesisStreamName: string;
  // Data S3 Bucket name
  private s3BucketName: string;

  constructor(scope: Construct, id: string, props: DataStreamConstructProps) {
    super(scope, id);

    const kinesisStreamsToKinesisFirehoseToS3 = new KinesisStreamsToKinesisFirehoseToS3(this, 'DataStream', {
      existingLoggingBucketObj: props.s3LoggingBucket,
      bucketProps: {
        serverAccessLogsPrefix: 'm2c2data/'
      }
    });
    this.kinesisStreamName = kinesisStreamsToKinesisFirehoseToS3.kinesisStream.streamName;
    this.s3BucketName = kinesisStreamsToKinesisFirehoseToS3.s3Bucket!.bucketName;
  }

  /**
   * Get Kinesis Data Stream name.
   * @return {string} Kinesis Data Stream name
   */
  getKinesisStreamName(): string {
    return this.kinesisStreamName;
  }

  /**
   * Get data S3 Bucket name.
   * @return {string} S3 Bucket name
   */
  getS3BucketName(): string {
    return this.s3BucketName;
  }
}