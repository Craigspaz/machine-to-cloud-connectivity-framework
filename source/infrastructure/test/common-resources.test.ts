// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import '@aws-cdk/assert/jest';
import { SynthUtils } from '@aws-cdk/assert';
import { Stack } from '@aws-cdk/core';
import { CommonResourcesConstruct } from '../lib/common-resources';

test('M2C2 common resources snapshot and private variables', () => {
  const stack = new Stack();
  const commonResources = new CommonResourcesConstruct(stack, 'TestCommonResources');

  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  expect(commonResources.getCloudWatchLogsPolicy()).toBeDefined();
  expect(commonResources.getS3LoggingBucket()).toBeDefined();
  expect(commonResources.getS3Bucket()).toBeDefined();
  expect(commonResources.getS3BucketName()).toBeDefined();
});