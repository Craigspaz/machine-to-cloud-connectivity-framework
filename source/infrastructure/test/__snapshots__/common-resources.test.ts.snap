// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`M2C2 common resources snapshot and private variables 1`] = `
Object {
  "Resources": Object {
    "TestCommonResourcesLogBucket77572DDE": Object {
      "DeletionPolicy": "Retain",
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W35",
              "reason": "This bucket is to store S3 logs, so it does not require access logs.",
            },
            Object {
              "id": "W51",
              "reason": "This bucket is to store S3 logs, so it does not require S3 policy.",
            },
          ],
        },
      },
      "Properties": Object {
        "AccessControl": "LogDeliveryWrite",
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "TestCommonResourcesM2C2Bucket2C991569": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "LoggingConfiguration": Object {
          "DestinationBucketName": Object {
            "Ref": "TestCommonResourcesLogBucket77572DDE",
          },
          "LogFilePrefix": "m2c2/",
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "TestCommonResourcesM2C2BucketPolicy147D6C50": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "TestCommonResourcesM2C2Bucket2C991569",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject",
                "s3:PutObject",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "TestCommonResourcesM2C2Bucket2C991569",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
}
`;
